### ISSUE TITLE
Improve cross-machine reproducibility, quality, and speed of national weight optimization by reformulating as QP and switching to scipy L-BFGS-B

### ISSUE THREAD

**Issue by donboyd5** [PSLmodels/tax-microdata-benchmarking#413]
@martinholmer, any comments on the following?

Prepared with the help of Claude code.

## Proposal summary

We can improve reproducibility, quality, and speed of national weight optimization by reformulating the problem as a quadratic program (QP) and switching the reweighting solver from PyTorch L-BFGS to scipy L-BFGS-B. We can do this in a non-disruptive manner by waiting until we update optimization targets to 2022 values. We can have a belt and suspenders by pushing a set of gold-standard weights (e.g., s006.csv or something similar) to master with each release, created on a contributor machine.

## Problem

Reweighting is not reproducible across machines. Two different computers — even two different CPUs — can produce materially different weights (`np.allclose()` fails). The most dramatic case is GPU vs CPU, but the problem affects any two environments with different floating-point behavior (different CPU architectures, different BLAS libraries, different OS/compiler combinations).

The root cause is structural: PyTorch L-BFGS uses `torch.clamp()` to enforce bound constraints, which creates flat gradient regions where the optimizer cannot distinguish between solutions. On this flat loss surface, floating-point non-associativity in `.sum()` reduction operations causes different hardware to push records to opposite bound extremes (e.g., multiplier 0.1 on one machine, 10.0 on another for the same record).

## Key Insight: The Existing Objective Is Already a Convex QP

The reweighting objective function in its original (NLP) formulation is:

$$\min_m \sum_j \left( \frac{\sum_i w_{0,i} \, m_i \, A_{ij}}{t_j + 1} - \frac{t_j}{t_j + 1} \right)^2 + \sum_i \lambda_i (m_i - 1)^2 \quad \text{s.t.} \quad m_{\min} \le m_i \le m_{\max}$$

where $m_i$ are weight multipliers, $w_{0,i}$ are original weights, $A_{ij}$ is the output matrix, $t_j$ are SOI targets, and $\lambda_i = \text{penalty} \cdot L_0 \cdot w_{0,i}^2 / \sum w_0^2$ is the per-record regularization weight (with $L_0$ = initial unpenalized loss).

The equivalent QP form is:

$$\min_m \frac{1}{2} m^T P m + q^T m \quad \text{s.t.} \quad m_{\min} \le m_i \le m_{\max}$$

where $P = 2(B^T B + \text{diag}(\lambda))$ and $q = -2(B^T c + \lambda)$, with $B_{ij} = w_{0,i} A_{ij} / (t_j + 1)$ and $c_j = t_j / (t_j + 1)$.

These are algebraically identical — expanding the squares in the NLP form and collecting terms gives the QP form. The NLP form is how we think about the problem; the QP form is what lets us exploit structure (analytical gradient, projected-gradient bounds, uniqueness guarantee).

This is a **bound-constrained convex quadratic program (QP)**. The Hessian is positive definite, which guarantees a **unique global minimum**. We have been solving this problem all along — but but because of the PyTorch L-BFGS characteristics below, we were not exploiting its QP structure:

- **Clamping instead of projected gradients**: `torch.clamp()` zeros out gradients at bounds, creating flat regions where the optimizer stalls. A proper projected-gradient method (like L-BFGS-B) projects the gradient onto the feasible set, maintaining useful gradient information at bounds.
- **Autograd instead of analytical gradient**: The gradient of a QP is a simple linear function. PyTorch's autograd computes it correctly but introduces unnecessary overhead and floating-point variability compared to the closed-form expression.
- **No uniqueness guarantee exploited**: Because clamping creates equivalent-loss plateaus at bounds, the PyTorch solver can settle at different points on the plateau depending on hardware-specific floating-point behavior. A solver that properly handles bounds converges to the unique minimum regardless of platform.

### Part 1: Reformulate the problem as a QP and switch to L-BFGS-B

Replace the PyTorch L-BFGS solver with **scipy's `L-BFGS-B`** (Fortran implementation), which properly exploits the QP structure (the "-B" means "bounds", which L-BFGS-B handles within the algorithm):

1. **Projected-gradient bounds** — no clamping artifacts, no flat gradient regions
2. **Analytical gradient** — closed-form QP gradient, deterministic and efficient:
   ```
   grad_i = 2 * sum_j B_ij * (sum_k B_kj * m_k - c_j) + 2 * lam_i * (m_i - 1)
   ```
   where `B_ij = w0_i * A_ij / (t_j + 1)` and `c_j = t_j / (t_j + 1)`
3. **Same objective function** — no change to what we're optimizing, only how
4. **Unique minimum convergence** — any machine converging tightly enough must find the same answer

The scipy L-BFGS-B solver has not been written to work on a GPU, but we don't need it to because it is so fast on a CPU.

#### Experimental Results

Tested on branch `experiment-grad-norm` (preserved on fork):

| Solver | Loss | Time | Converged | Cross-machine reproducible |
|--------|------|------|-----------|---------------------------|
| PyTorch L-BFGS (GPU) | 0.1126 | ~90s | Stagnated | No |
| PyTorch L-BFGS (CPU) | 0.1126 | ~180s | Stagnated | No |
| **Scipy L-BFGS-B** | **0.1065** | **422s** | **Yes (grad < 1e-5)** | **Yes**\* |
| OSQP (ADMM) | 0.1057 | 283s | No (50K iter limit) | Yes\* |

\*Deterministic given same BLAS library

Scipy L-BFGS-B achieves **better loss** (0.1065 vs 0.1126) because it can properly optimize near bounds rather than stalling on clamping plateaus. Runtime of ~7 minutes is acceptable for a build step that runs once per release.

OSQP (a dedicated QP solver using ADMM) was also tested but could not converge to tight tolerances within reasonable time — a fundamental limitation of first-order ADMM methods for high-accuracy solutions.

#### What Changes

- **Solver**: PyTorch L-BFGS → scipy L-BFGS-B (Fortran)
- **Gradient**: autograd → analytical (closed-form)
- **Bounds handling**: `torch.clamp()` → projected gradients (L-BFGS-B native)
- **Objective function**: unchanged
- **GPU dependency**: eliminated (CPU-only, deterministic)

#### What Doesn't Change

- The objective function and its meaning
- The targets, penalties, and bound constraints
- The output format (weight multipliers applied to s006)

The branch `experiment-grad-norm` on the fork contains the working scipy implementation and all experimental logs.

### Part 2: Create gold-star weights and include them in master branch

In addition to the above, we can create a set of gold-star weights -- optimum weights produced on a contributor's machine. We can commit them to master. Users can check whether their weights match the gold-star weights within np.allclose() defaults, or they can just use the gold-star weights directly.

**Comment by martinholmer** [comment]
@donboyd5 proposed a new reweighting algorithm in [this comment](https://github.com/PSLmodels/tax-microdata-benchmarking/issues/413#issue-3974764870) in issue #413.

@donboyd5, my **first question** after reading you comment is "Why didn't Claude Code suggest this approach from the beginning?"   Claude can see the TMD repo is already using the `scipy` `L-BFGS-B` optimization method for other things, so why didn't Claude suggest this approach from the beginning?   And more to the point, will Claude suggest a completely different approach next time you ask for advice?

My **second question** is that you say:
```
Unique minimum convergence — any machine converging tightly enough must find the same answer
```
But then say in Part 2 that there is a need for "gold-star weights" to be included in the repo.
Why is there a need for "gold-star weights" when "any machine converging tightly enough 
must find the same answer"?

My **third question** is why is the proposed optimization method so slow?   Did you use `jax` to provide the method with gradient info like I did in the area weights optimization?

And my **fourth question** is where is your `experiment-grad-norm` branch --- I don't see it on GitHub.

**Comment by donboyd5** [comment]
https://github.com/donboyd5/tax-microdata-benchmarking/tree/experiment-grad-norm

**Comment by donboyd5** [comment]
I said we can create gold-star weights, not that we have to. It's one way to guarantee a user can hit given results. Anyway, I'll address this when I address 2022.
